binding.c

```diff
@@ -227,6 +263,278 @@ bare_dns_lookup(js_env_t *env, js_callback_info_t *info) {
+
+    struct ares_txt_ext *reply;
+
+    status = ares_parse_txt_reply_ext(buf, len, &reply);
+    assert(status == ARES_SUCCESS);
+
+    err = js_get_null(env, &args[0]);
+    assert(err == 0);
+
+    js_value_t *result;
+    err = js_create_array(env, &result);
+    assert(err == 0);
+
+    uint32_t i = 0;
+
+    for (struct ares_txt_ext *next = reply; next != NULL; next = next->next) {
+      js_value_t *chunk;
+      err = js_create_string_utf8(env, (utf8_t *) next->txt, next->length, &chunk);
+      assert(err == 0);
+
+      if (next->record_start) {
+        js_value_t *record;
+        err = js_create_array(env, &record);
+        assert(err == 0);
+
+        err = js_set_element(env, record, 0, chunk);
+        assert(err == 0);
+
+        err = js_set_element(env, result, i++, record);
+        assert(err == 0);
+      } else {
+        js_value_t *record;
+        err = js_get_element(env, result, i, &record);
+        assert(err == 0);
+
+        uint32_t len;
+        err = js_get_array_length(env, record, &len);
+        assert(err == 0);
+
+        err = js_set_element(env, record, len, chunk);
+        assert(err == 0);
+      }
+    }
+
+    args[1] = result;
+
+    ares_free_data(reply);
+  } else {
+    js_value_t *message;
+    err = js_create_string_utf8(env, (utf8_t *) ares_strerror(status), -1, &message);
+    assert(err == 0);
+
+    err = js_create_error(env, NULL, message, &args[0]);
+    assert(err == 0);
+
+    err = js_get_null(env, &args[1]);
+    assert(err == 0);
+  }
+
+  js_call_function(env, ctx, cb, 2, args, NULL);
+
+  err = js_close_handle_scope(env, scope);
+  assert(err == 0);
+}
+
+static js_value_t *
+bare_dns_resolve_txt(js_env_t *env, js_callback_info_t *info) {
+  int err;
+
+  size_t argc = 4;
+  js_value_t *argv[4];
+
+  err = js_get_callback_info(env, info, &argc, argv, NULL, NULL);
+  assert(err == 0);
+
+  assert(argc == 4);
+
+  bare_dns_resolver_t *resolver;
+  err = js_get_arraybuffer_info(env, argv[0], (void **) &resolver, NULL);
+  assert(err == 0);
+
+  size_t len;
+  err = js_get_value_string_utf8(env, argv[1], NULL, 0, &len);
+  assert(err == 0);
+
+  len += 1 /* NULL */;
+
+  utf8_t *hostname = malloc(len);
+  err = js_get_value_string_utf8(env, argv[1], hostname, len, NULL);
+  assert(err == 0);
+
+  js_value_t *handle;
+
+  bare_dns_resolve_req_t *req;
+  err = js_create_arraybuffer(env, sizeof(bare_dns_resolve_req_t), (void **) &req, &handle);
+  assert(err == 0);
+
+  req->resolver = resolver;
+  req->env = env;
+
+  err = js_create_reference(env, argv[2], 1, &req->cb);
+  assert(err == 0);
+
+  err = js_create_reference(env, argv[3], 1, &req->ctx);
+  assert(err == 0);
+
+  ares_status_t status;
+  status = ares_query_dnsrec(req->resolver->channel, (char *) hostname, ARES_CLASS_IN, ARES_REC_TYPE_TXT, bare_dns__on_resolve_txt, req, NULL);
+
+  if (err != ARES_SUCCESS) {
+    js_throw_error(env, NULL, ares_strerror(err));
+    return NULL;
+  }
+
+  free(hostname);
```

Comment:
```suggestion
  free(hostname);

  if (err != ARES_SUCCESS) {
    js_throw_error(env, NULL, ares_strerror(err));
    return NULL;
  }
```

Otherwise the memory will leak on error.

Reply:
Fixed at: https://github.com/holepunchto/bare-dns/pull/3/commits/780ff539e060de34397d0f7f1cc0da8f640888f5